### PR TITLE
Sample set and annotation improvements for SVConcordance

### PR DESCRIPTION
**Pull request by mwalker174**
- Relaxes restrictions for allowed samples in SVConcordance: the tool can now accept eval/truth VCFs with arbitrary sample sets and will have genotype concordance metrics computed on the intersection of the sample sets. All available samples are still used for AF/AC annotations. Integration tests added for cases when the samples sets are overlapping but not equal.
- Small additional improvements for sites-only VCFs: concordance annotations will now be `.` instead of `NaN` for example. Integration test added for this case.
- Improved behavior for eval AF annotations: these will not be recalculated if they already exist.
- Improved behavior for truth AF annotations: these will now only be recalculated if they don't exist in the input truth VCF.
- Updated tool doc

### CODE REVIEW

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/sv/SVConcordanceIntegrationTest.java [hunk 1]
```diff
@@ -227,4 +227,152 @@ public void testSelf() {
             }
         }
     }
+
+    @Test
+    public void testSelfTruthSubset() {
```
Comment (ldgauthier) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/sv/SVConcordanceIntegrationTest.java:232]: Very elegant

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/sv/SVConcordanceIntegrationTest.java [hunk 2]
```diff
@@ -227,4 +227,152 @@ public void testSelf() {
...
+                    MathUtilsUnitTest.assertEqualsDoubleSmart(outputVariant.getAttributeAsDouble(GATKSVVCFConstants.COPY_NUMBER_CONCORDANCE_INFO, -1.), 1.0, TOLERANCE);
+                } else {
+                    MathUtilsUnitTest.assertEqualsDoubleSmart(outputVariant.getAttributeAsDouble(GATKSVVCFConstants.GENOTYPE_CONCORDANCE_INFO, -1.), 1.0, TOLERANCE);
+                }
+            }
+        }
+    }
+
+    @Test
+    public void testSelfEvalSubset() {
```
Comment (ldgauthier) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/sv/SVConcordanceIntegrationTest.java:291]: There's a lot of duplicate code here, so I might be inclined to suggest a data provider, but it's also sort of a weird case where we would have a genericified test that expects perfect concordance.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/sv/SVConcordanceIntegrationTest.java [hunk 3]
```diff
@@ -227,4 +227,152 @@ public void testSelf() {
...
+                .add(AbstractConcordanceWalker.TRUTH_VARIANTS_LONG_NAME, truthVcfPath)
+                .add(AbstractConcordanceWalker.EVAL_VARIANTS_SHORT_NAME, evalVcfPath);
+
+
+        runCommandLine(args, SVConcordance.class.getSimpleName());
+
+        final Pair<VCFHeader, List<VariantContext>> outputVcf = VariantContextTestUtils.readEntireVCFIntoMemory(output.getAbsolutePath());
+        final List<SVCallRecord> inputEvalVariants = VariantContextTestUtils.readEntireVCFIntoMemory(evalVcfPath).getValue()
+                .stream().map(SVCallRecordUtils::create).collect(Collectors.toList());
+        Assert.assertEquals(outputVcf.getValue().size(), inputEvalVariants.size());
```
Comment (ldgauthier) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/sv/SVConcordanceIntegrationTest.java:376]: This seems a little lax.  Are there new annotations you expect in the output?

**Review by ldgauthier — APPROVED** [review]
Mostly questions for my own edification.

The sites-only test seems weak, but honestly I didn't go back to see what the expected behavior is.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/sv/SVConcordance.java [hunk 4]
```diff
@@ -42,11 +42,12 @@
...
- * genotype concordance metrics, including allele frequency of the truth variant. See output header for descriptions
- * of the specific fields. Note that genotypes of samples that are present in the evaluation VCF but not the truth
- * VCF are assumed to be no-calls for concordance. For multi-allelic CNVs, only a copy state concordance metric is
- * annotated.
+ * after meeting minimum overlap criteria. Evaluation VCF variants that are sucessfully matched are annotated with
+ * genotype concordance metrics, including allele frequency of the truth variant. Concordance metrics are computed
+ * on the intersection of sample sets of the two VCFs, but all other annotations including variant truth status
+ * and allele frequency use all records and samples available. See output header for descriptions
+ * of the specific fields. For multi-allelic CNVs, only a copy state concordance metric is
+ * annotated. Allele frequencies will be recalculated automatically if unavailable in the provided VCFs.
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/sv/SVConcordance.java:50]: No but that's a good idea for the future. Right now we're using this for filtering so the important thing is the vcf annotations.

File: src/test/java/org/broadinstitute/hellbender/tools/sv/SVTestUtils.java [hunk 5]
```diff
@@ -388,14 +395,6 @@ public static final Map<String, Object> keyValueArraysToMap(final String[] keys,
         return map;
     }
 
-    // Note that strands may not be valid
-    public static SVCallRecord newCallRecordWithLengthAndTypeAndChrom2(final Integer length, final GATKSVVCFConstants.StructuralVariantAnnotationType svtype, final String chrom2) {
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/test/java/org/broadinstitute/hellbender/tools/sv/SVTestUtils.java:392]: I can't remember exactly when/where this was used, but I noticed it's unused now. I may have needed correct strands for new tests in the last PR and just used a more general method.

File: src/main/java/org/broadinstitute/hellbender/tools/sv/concordance/SVConcordanceAnnotator.java [hunk 6]
```diff
@@ -76,54 +86,66 @@ public SVCallRecord annotate(final ClosestSVFinder.ClosestPair pair) {
...
-            attributes.put(GATKSVVCFConstants.GENOTYPE_CONCORDANCE_INFO, metrics.GENOTYPE_CONCORDANCE);
-            attributes.put(GATKSVVCFConstants.NON_REF_GENOTYPE_CONCORDANCE_INFO, metrics.NON_REF_GENOTYPE_CONCORDANCE);
-            attributes.put(GATKSVVCFConstants.HET_PPV_INFO, metrics.HET_PPV);
-            attributes.put(GATKSVVCFConstants.HET_SENSITIVITY_INFO, metrics.HET_SENSITIVITY);
-            attributes.put(GATKSVVCFConstants.HOMVAR_PPV_INFO, metrics.HOMVAR_PPV);
-            attributes.put(GATKSVVCFConstants.HOMVAR_SENSITIVITY_INFO, metrics.HOMVAR_SENSITIVITY);
-            attributes.put(GATKSVVCFConstants.VAR_PPV_INFO, metrics.VAR_PPV);
-            attributes.put(GATKSVVCFConstants.VAR_SENSITIVITY_INFO, metrics.VAR_SENSITIVITY);
-            attributes.put(GATKSVVCFConstants.VAR_SPECIFICITY_INFO, metrics.VAR_SPECIFICITY);
+            attributes.put(GATKSVVCFConstants.GENOTYPE_CONCORDANCE_INFO, Double.isNaN(metrics.GENOTYPE_CONCORDANCE) ? null : metrics.GENOTYPE_CONCORDANCE);
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/sv/concordance/SVConcordanceAnnotator.java:93]: Yeah, I think pysam might crash with those too.

File: src/main/java/org/broadinstitute/hellbender/tools/sv/concordance/SVConcordanceAnnotator.java [hunk 7]
```diff
@@ -76,54 +86,66 @@ public SVCallRecord annotate(final ClosestSVFinder.ClosestPair pair) {
...
             }
         }
 
         return SVCallRecordUtils.copyCallWithNewAttributes(recordWithGenotypes, attributes);
     }
 
+    private boolean hasAlleleFrequencyAnnotations(final SVCallRecord record) {
+        Utils.nonNull(record);
+        final Map<String, Object> truthAttr = record.getAttributes();
+        return (truthAttr.containsKey(GATKSVVCFConstants.TRUTH_ALLELE_COUNT_INFO) && truthAttr.get(GATKSVVCFConstants.TRUTH_ALLELE_COUNT_INFO) != null)
```
[the remark this answers is not in the document]
Comment (mwalker174) [reply src/main/java/org/broadinstitute/hellbender/tools/sv/concordance/SVConcordanceAnnotator.java:144]: Good catch! Should be looking for regular AC/AF/AN here. I've fixed this and added regression tests.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/sv/SVConcordanceIntegrationTest.java (hunk 2 above)
Comment (mwalker174) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/sv/SVConcordanceIntegrationTest.java:291]: Okay I've created a `assertPerfectConcordance()` to do the checking at least.

**Pull request merged**